lib/spack/spack/binary_distribution.py

```diff
@@ -0,0 +1,260 @@
+            if spack.relocate.needs_binary_relocation(filetype):
+                rel_path_name = os.path.relpath(path_name, spec.prefix)
+                binary_to_relocate.append(rel_path_name)
+            elif spack.relocate.needs_text_relocation(filetype):
+                rel_path_name = os.path.relpath(path_name, spec.prefix)
+                text_to_relocate.append(rel_path_name)
+
+    # Create buildinfo data and write it to disk
+    buildinfo = {}
+    buildinfo['buildpath'] = spack.store.layout.root
+    buildinfo['relocate_textfiles'] = text_to_relocate
+    buildinfo['relocate_binaries'] = binary_to_relocate
+    filename = buildinfo_file_name(spec)
+    with open(filename, 'w') as outfile:
+        outfile.write(yaml.dump(buildinfo, default_flow_style=True))
+
+
+def tarball_directory_name(spec):
+    """
+    Return name of the tarball directory according to the convention
+    <os>-<architecture>/<compiler>/<package>/
+    """
+    return "%s/%s/%s" % (get_full_system_from_platform(),
+                         str(spec.compiler).replace("@", "-"),
+                         spec.name)
+
+
+def tarball_name(spec, ext):
+    """
+    Return the name of the tarfile according to the convention
+    <os>-<architecture>-<package>-<dag_hash><ext>
+    """
+    return "%s-%s-%s-%s%s" % (get_full_system_from_platform(),
+                            spec.name,
+                            spec.version,
+                            spec.dag_hash(),
+                            ext)
+
+
+def tarball_path_name(spec, ext):
+    """
+    Return the full path+name for a given spec according to the convention
+    <tarball_directory_name>/<tarball_name>
+    """
+    return os.path.join(tarball_directory_name(spec),
+                        tarball_name(spec, ext))
+
+
+def build_tarball(spec, outdir, force=False, key=None):
+    """
+    Build a tarball from given spec and put it into the directory structure
+    used at the mirror (following <tarball_directory_name>).
+    """
+    tarfile_name = tarball_name(spec, '.tar.gz')
+    tarfile_dir = join_path(outdir, tarball_directory_name(spec))
+    tarfile_path = join_path(tarfile_dir,tarfile_name)
+    mkdirp(tarfile_dir)
+    if os.path.exists(tarfile_path):
+        if force:
+            os.remove(tarfile_path)
+        else:
+            tty.warn("file exists, use -f to force overwrite: %s" % tarfile)
+            return
+
+    spec_file = join_path(spec.prefix, ".spack", "spec.yaml")
+
+    # create info for later relocation and create tar
+    write_buildinfo_file(spec)
+    with tarfile.open(tarfile_path, 'w:gz') as tar:
+        tar.add(name = '%s' % spec.prefix 
+               , arcname = '%s' % os.path.basename(spec.prefix))
+
+    # Sign the packages.
+    #spack gpg sign [--key key] tarfile_path
+    #spack gpg sign [--key key] tarfile_path + '/spec.yaml'
+    path1='%s.asc' % tarfile_path
+    with open(path1,'a'):
+        os.utime(path1, None)       
```

Comment:
Why is this required? Or is this just in lieu of `spack gpg` not being available yet?